### ISSUE TITLE
[ICL] occasional display audio codec probe failure - fail to get afg sub nodes

### ISSUE THREAD

**Issue by kv2019i** [thesofproject/linux#1184]
This is a fork of bug https://github.com/thesofproject/linux/issues/1128

This issues happens occasionally on at least ICL system. The main fingerprint of the bug are following error messages in dmesg:

Sep  6 13:58:48 cflsdp2019 kernel: [  803.216002] snd_hda_codec_hdmi ehdaudio0D2: HDMI: failed to get afg sub nodes
Sep  6 13:58:48 cflsdp2019 kernel: [  803.216004] snd_hda_codec_hdmi ehdaudio0D2: patch failed -22
Sep  6 13:58:48 cflsdp2019 kernel: [  803.216084] snd_hda_codec_hdmi ehdaudio0D2: ASoC: failed to probe component -22
Sep  6 13:58:48 cflsdp2019 kernel: [  803.216103] skl_hda_dsp_generic skl_hda_dsp_generic: ASoC: failed to instantiate card -22

When this bug triggers, the card probe fails and "aplay -l" doesn't show any PCM devices for SOF.

**Comment by kv2019i** [comment]
FYI @fredoh9 @ClarexZhou

**Comment by kv2019i** [comment]
Some evidence for the explanation that delays in graphics (either via HDMI detection failures or by increase DRM debugs) cause this error to happen. I.e. this could be a timing race in probe actions. Still looking for more reliable steps to reproduce.

**Comment by ranj063** [comment]
@kv2019i the chrome validation team have notified that they have encountered this problem in CML with a very low reproduction rate as well.

**Comment by sathyap-chrome** [comment]
@ranj063  THanks. @kv2019i  we are seeing this once every 1000 - 2000 reboots.
Attached the dmesg for reference.
 I see some backtraces in gpu - not sure if they are related.
[dmesg_hdmi_failed_error_log.txt](https://github.com/thesofproject/linux/files/3615192/dmesg_hdmi_failed_error_log.txt)
[vblank_error_backtrace.txt](https://github.com/thesofproject/linux/files/3615193/vblank_error_backtrace.txt)

**Comment by kv2019i** [comment]
@sathyap-chrome wrote:
> @ranj063 THanks. @kv2019i we are seeing this once every 1000 - 2000 reboots.
> Attached the dmesg for reference.

Thanks for the log! Adding CML label for this bug as well. 

Your log shows one difference. It seems the HDMI codec gives an error already at probe time:

[    3.638663] sof-audio-pci 0000:00:1f.3: hda codecs found, mask 4
[    3.647168] sof-audio-pci 0000:00:1f.3: CORB reset timeout#1, CORBRP = 0

... in the logs we have from other systems, this "reset timeout" is not visible. But keeping in this same bug for now, could be still the same rootcause.

For now focus is to find an easier way to reproduce. If get further logs, please include to this bug.

**Comment by kv2019i** [comment]
Today's results: reboot tests with 0/100 fails on CFL HDA system. Will continue with longer test runs and expand test to other systems.

**Comment by ranj063** [comment]
> [ 3.647168] sof-audio-pci 0000:00:1f.3: CORB reset timeout#1, CORBRP = 0

@kv2019i back when we were fixing the IPC timeouts on CML/CFL , Im pretty sure we saw the same timeout. But I cant seem to recollect how this got resolved. Do you remember?

**Comment by kv2019i** [comment]
> Today's results: reboot tests with 0/100 fails on CFL HDA system. Will continue with longer test runs and expand test to other systems.

Upto 500 boots on this CFL system and no failures. I will switch to other platforms and try to reproduce. @YvonneYang2 @ClarexZhou have got this much more frequently, I'll try to repeat your steps.

**Comment by kv2019i** [comment]
Daily update: I now reproduce this with 100% reproduction rate on one ICL system with following properties:
 - ICL system that has a long delay in HDA codec modprobe (not yet rootcaused, but this delay seems to be key in triggering this bug),
 - no HDMI or DP cables plugged in when the system is booted

The sequence of events that happens is that there is a notable delay between (a) SOF hda.c doing the HDA codec probe (and detecting a i915 HDA audio codec on the system ) and (b) codec driver probe being run. 

The cause of problems seems to be i915 display driver going to low power state between (a) and (b). 
If I prevent low-power states, the problem goes away. Also when I modify the sequence so that (a) and (b) are close enough together that i915 does disable the related power well, problem disappears. 
With legacy driver, the module probe mechanism is different, so the bug is not triggered.

I think similar sequence can happen on other platforms if the system boot timing is for some reason modified occasionally (or with specific external conditions, like which monitors are connected, and so forth). This could explain seeing the error with very low reproduction rates.

I don't have a candidate fix at the moment. The power-up sequences we do at (b) look correct, and i915 driver logs seem correct. So spec-wise everything should be ok, despite the timing variation and autosuspend kicking in during boot for some i915 power domains.

Now I at least have reproducible test case, so will continue debug with this.

**Comment by kv2019i** [comment]
It seems similar issue as been reported earlier against i915:
https://bugs.freedesktop.org/show_bug.cgi?id=111214

**Comment by kv2019i** [comment]
Update: the scenario with 100% reproduction rate turned out to be a different bug. So good news is that that this unblocks upstreaming of the HDA-HDMI rework (fixed series at https://github.com/thesofproject/linux/pull/1155 -> testing ongoing ). Bad news is that the original problem of occasional probe failures may still be present. We have been discussing this today with graphics driver folks and there's one potential fix to display driver code that may help (pending results from gfx CI). Keeping this open for now.

**Comment by kv2019i** [comment]
Update: a new scenario discovered where probe error can be triggered:

- boot system (tested on ICL)
- unload audio drivers
- put system to S3 suspend (e.g. with rtcwake)
- resume 
- load audio drivers
-> HDMI probe fails with high reproduction rate and display driver logs show that the codec does not come out of sleep state (leading to probe error with both legacy and SOF drivers)

**Comment by ranj063** [comment]
@kv2019i I was experimenting with the i915 init sequence and codec probe with my MFD PR and I can easily trigger the codec probe error with it. I am now convinced that this is a timing issue with how/when the i915 and HDA controller are initialized and the codecs are probed.

**Comment by kv2019i** [comment]
Current understanding (and solutions) for the bug:

CML issue -> this is fixed by Takashi Iwai's patch in upstream (2756d9143aa5 ALSA: hda - Fix intermittent CORB/RIRB stall on Intel chips)

CFL issue (logs in original bug report) -> caused by a bug in new HDMI codec driver patch series, now fixed
 
GLK/ICL issue -> fix candidate submitted to i915 driver -> Potential bugfix to this issue submitted on intel-gfx list:
https://lists.freedesktop.org/archives/intel-gfx/2019-October/214621.html
(update, backported to SOF tree on top of https://github.com/thesofproject/linux/pull/1155 .. fyi @ranj063 )

So if all goes well, we have a solution to all known instances of this problem. The ICL/GLK fix is trickiest and we need a lot more validation before we can conclude it is fixed by current set of patches.

**Comment by kv2019i** [comment]
Seen on TGL as well, assigning P2 for now.

**Comment by kv2019i** [comment]
Patches to gen10+ display have been accepted upstream -> https://patchwork.freedesktop.org/series/67528/

New issue found in suspend/reload combined stress test that affects only SOF with patch_hdmi driver -> patch sent to alsa-devel today and backprot include in patch series #1155 

If testing is ok for #1155 , this bug can be closed. All known issues fixed.

**Comment by kv2019i** [comment]
Current status: all known issues fixed for ICL and other platforms, but it seems the issue is still hit sometimes on GLK in CI, even with all the i915 and adsp patches. Here're three recent test results (random patches submitted to CI yesterday and today):

https://sof-ci.01.org/linuxpr/PR1290/build2066/devicetest/
https://sof-ci.01.org/linuxpr/PR1295/build2071/devicetest/
https://sof-ci.01.org/linuxpr/PR1300/build2072/devicetest/

In some cases tests pass on GLK Bobba but not all. The error hit is "HDMI HDA Codec ehdaudio0D2: HDMI: failed to get afg sub nodes" in verify_pcm_list, so fits this bugs fingerprint. Our sof-dev tip doesn't yet have all the display driver fixes, but unfortunately the GLK errors happen also with PR1155 to which I have backported all the display patches:

https://sof-ci.01.org/linuxpr/PR1155/build2045/devicetest/

**Comment by wenqingfu** [comment]
> Patches to gen10+ display have been accepted upstream -> https://patchwork.freedesktop.org/series/67528/

@kv2019i are these patch cherry picked to sof-dev now? which PR contains it?

**Comment by kv2019i** [comment]
@wenqingfu wrote:
>> Patches to gen10+ display have been accepted upstream -> https://patchwork.freedesktop.org/series/67528/
> 
> @kv2019i are these patch cherry picked to sof-dev now? which PR contains it?

@wenqingfu They have been part of PR1155, but I now made a clean PR so the patches can be picked to sof-dev: https://github.com/thesofproject/linux/pull/1304
... this time, CI only had one pause-resume fail for PR1304 (e.g. no HDMI failures).

**Comment by kv2019i** [comment]
In CI stats for 11th Oct so far (3 PRs with devicetest round), no probe error on GLK:

https://sof-ci.01.org/linuxpr/PR1301/build2086/devicetest/
https://sof-ci.01.org/linuxpr/PR1300/build2084/devicetest/
https://sof-ci.01.org/linuxpr/PR1155/build2085/devicetest/

**Comment by wenqingfu** [comment]
From last several days' testing, the GLK issue can now be confirmed as a device setup issue. 

I think this issue can be closed, but leave to @kv2019i for confirm and close

**Comment by kv2019i** [comment]
Ack @wenqingfu . Let's open new bugs for any new findings.
